File: all-paths-from-source-to-target/solution-recursive.js
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */
var allPathsSourceTarget = function(graph) {
    var path = [];
    var paths = [];
    getPaths(graph, 0, path, paths);
    return paths; 
};

function getPaths(graph, index, path, paths) {
    var path = path.slice(); 
    path.push(index)
    
    if (index == (graph.length - 1)) {
        paths.push(path);
        return; 
    };

    graph[index].forEach(function(i) { 
        getPaths(graph, i, path, paths)
    });
}; 

File: all-paths-from-source-to-target/solution-recursive.py
class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """
        paths = []
        path = []
        iterate(0, graph, path, paths)
        return paths
        
def iterate(i, graph, path, paths): 
    newPath = path[:]
    newPath.append(i)
    
    if i == len(graph) - 1: 
        paths.append(newPath)
    else: 
        for node in graph[i]:
            iterate(node, graph, newPath, paths)
        
